### PR TITLE
Return groups from the database

### LINKED ISSUE #21: Attach files and notes to machines in the admin panel
When setting up/editing a machine in the admin panel, allow for admins to make notes (eg. noisy GPU fan) or upload documents (eg. invoices) that are attached to that particular machine.

Should probably be stored locally (don't want to transmit them elsewhere)

### LINKED ISSUE #76: Support Workstation Groups through the Back-End

### LINKED ISSUE #124: Fix groups

### PR DESCRIPTION
**Pull request by danielg0**
In order to return groups from the database, make a major change to the database interface. For "output" functions, rather than returning the uplink types (those for satellite-to-groundstation communication), return the broadcast types (those for webapi-to-frontend communication). This means we can remove all the zipping work from the front-end.

Closes #124
Closes #76
Makes progress toward closing #21 (we still want to be able to attach files)

### CODE REVIEW

**Comment by danielg0** [comment]
Please do review if you find something wrong, but I've left it as draft for the moment because I want to:

- [x] Double check all the tests are working
- [x] Write some new tests (covering group changes propagating to the db, etc.)
- [x] Replace some of the very long comparisons with reflection
- [x] Get confirmation from @NathanielB123 that this will work right with the json the frontend's expecting

File: internal/database/interface.go
Comment (aDotInTheVoid) [inline internal/database/interface.go]: NIT: Can this be `const` instead?

Comment (danielg0) [reply internal/database/interface.go]: Change made in eb0287e

**Comment by NathanielB123** [comment]
> Please do review if you find something wrong, but I've left it as draft for the moment because I want to:
> 
> * [x]  Double check all the tests are working
> * [x]  Write some new tests (covering group changes propagating to the db, etc.)
> * [x]  Replace some of the very long comparisons with reflection
> * [x]  Get confirmation from @NathanielB123 that this will work right with the json the frontend's expecting

On last point - only differences I can spot are:
- front-end not using `uuid` (but extra fields are handled gracefully by default iirc so shouldn't be a problem) 
- on the front-end the `workstations` field is named (extremely inconsistently) `workStations`  - not sure if JSON marshalling is case sensitive, but honestly I'd be in support of renaming the field on the front-end with this PR.

**Comment by danielg0** [comment]
> * on the front-end the `workstations` field is named (extremely inconsistently) `workStations`  - not sure if JSON marshalling is case sensitive, but honestly I'd be in support of renaming the field on the front-end with this PR.

Done

**Comment by danielg0** [comment]
Acc, there's a conflict, I'm going to do that in a new PR

**Comment by danielg0** [comment]
> * on the front-end the `workstations` field is named (extremely inconsistently) `workStations`  - not sure if JSON marshalling is case sensitive, but honestly I'd be in support of renaming the field on the front-end with this PR.

See #148

File: internal/database/inmemory.go
Comment (Euphoride) [inline internal/database/inmemory.go:23]: Is this necessary? These side-effects won't be exposed, and inMemory is only intended to be used as a mock database. Control does not read what it writes here and thus I don't believe this is necessary.

File: internal/database/postgres.go
Comment (Euphoride) [inline internal/database/postgres.go:66]: Postgres should be using an extension to represent a UUID type here. We should also be using a Go UUID type, and the frontend should use a typescript UUID type.

File: internal/database/postgres.go
Comment (Euphoride) [inline internal/database/postgres.go:318]: With primarily Read queries, using a transaction is unnecessary and inefficient. There is no need to enforce read-write race condition freedom as we accept eventual consistency over to-the-instant consistency.

File: internal/database/postgres.go
Comment (Euphoride) [inline internal/database/postgres.go]: This is incorrect. The sql.ErrNoRows error is typically relevant when querying for a single row (e.g., with QueryRow). When iterating over multiple rows with Query, an empty result set doesn't result in an error but simply leads to no iterations of the loop.

File: internal/database/postgres.go
Comment (Euphoride) [inline internal/database/postgres.go:323]: NIT: Preallocating the size of this is an efficiency boost.

File: internal/database/postgres.go
Comment (Euphoride) [inline internal/database/postgres.go]: Use default values in the SQL query instead rather than custom logic.

File: internal/database/postgres.go
Comment (Euphoride) [inline internal/database/postgres.go:315]: Isn't it possible here to use SQL's grouping mechanism to group things for you. SQL engines are **heavily** optimised for this work, why are we reinventing the wheel?

File: internal/database/postgres.go
Comment (Euphoride) [inline internal/database/postgres.go:388]: Better aliases on this query would be ideal.

File: internal/database/postgres.go
Comment (Euphoride) [inline internal/database/postgres.go]: DITTO on `ErrNoRows`

File: internal/database/postgres.go
Comment (Euphoride) [inline internal/database/postgres.go]: DITTO on ErrNoRows

**Review by Euphoride — DISMISSED** [review]

File: internal/database/postgres.go
Comment (Euphoride) [inline internal/database/postgres.go:356]: NIT: Not ideal, pollutes the error returned if rollback also fails. Go wants simple error messages, and simple flow. I'd suggest logging the rollback error and returning `err` so that you can actually test the type/presence/etc of this error properly without a rollback failure getting in the way.

File: internal/database/postgres.go
Comment (Euphoride) [inline internal/database/postgres.go:362]: Bar me misunderstanding this code, this shouldn't have an inner loop that defers to a new SQL query. Rather, a clean, well-formatted and efficient SQL query should be constructed to defer this logic to the SQL Engine.

The SQL query execution planner has an almost certain chance of planning something more efficient than this, with much better performance.

File: internal/database/postgres.go
Comment (danielg0) [reply internal/database/postgres.go:66]: This isn't a change I want to include in this pr, see #129

File: internal/database/inmemory.go
Comment (danielg0) [reply internal/database/inmemory.go:23]: I believe so, these side effects are exposed in the `Workstation` struct on a call to `LatestData`

**Review by NathanielB123 — APPROVED** [review]

File: internal/database/postgres.go
Comment (danielg0) [reply internal/database/postgres.go]: Yep, fixed in 2f3d334

File: internal/database/postgres.go
Comment (danielg0) [reply internal/database/postgres.go]: Fixed in 2f3d334

File: internal/database/postgres.go
Comment (danielg0) [reply internal/database/postgres.go]: Fixed in 2f3d334

File: internal/database/postgres.go
Comment (danielg0) [reply internal/database/postgres.go:356]: I disagree that this pollutes the error. `errors.Join()` returns a wrapped error. If you're testing for the presence/type of an error, you should be using [`errors.Is()`](https://pkg.go.dev/errors#Is), it checks all the errors wrapped and returns true if any of the errors matched the target error.

File: internal/database/postgres.go
Comment (danielg0) [reply internal/database/postgres.go:388]: Just to be clear, the aliases you're referring to here are how `g` refers to the `Gpus` table and `s` refers to the `Stats` table?

File: internal/database/postgres.go
Comment (danielg0) [reply internal/database/postgres.go]: Done in f369989

File: internal/database/postgres.go
Comment (danielg0) [reply internal/database/postgres.go:323]: How would you suggest I get the number of groups in the database before we iterate through the returned rows - another SQL query before?

File: internal/database/postgres.go
Comment (danielg0) [reply internal/database/postgres.go:362]: Unless we find that our database queries are being too slow, I'd like to treat this as a non-blocking style issue. I think the existing query in `getGpus` is already too long to understand quickly, and really don't want to combine that with joining the `machine` table and pulling out all of its fields too. I personally prefer shorter queries that are more decomposed, because I find them easier to understand, debug and modify.

File: internal/database/postgres.go
Comment (danielg0) [reply internal/database/postgres.go:315]: I don't think a `GROUP BY` query, if that's what you're suggesting, would work for `LatestData`, because we don't want to aggregate the GPUs so there's one row per machine/per group

File: internal/database/postgres.go
Comment (aDotInTheVoid) [reply internal/database/postgres.go:362]: Yeah, I agree that it's fine to have kinda sucky SQL queries for now. We can always change them later, and theirs no reason holding up integration on this.

File: internal/database/postgres.go
Comment (aDotInTheVoid) [reply internal/database/postgres.go:318]: Eh, you could end up with some weird anomalies of machines/gpu's moving inside the same read. I don't think DB perf is a major concern, so I'm happy to merge this as is.

**Comment by aDotInTheVoid** [comment]
LGTM, when conflicts resolved.

**Pull request merged**
